_posts/2019-12-20-differential-privacy/differential_privacy.Rmd

```diff
@@ -0,0 +1,470 @@
+ds_train <- zip_datasets(x_train, y_train)
+ds_test <- zip_datasets(x_test, y_test)
+
+batch_size <- 32
+
+ds_train <- ds_train %>% 
+  dataset_shuffle(n_train) %>%
+  # dataset_repeat is needed because of pre-TF 2 style
+  # hopefully at a later time, the code can run eagerly and this is no longer needed
+  dataset_repeat() %>%
+  dataset_batch(batch_size, drop_remainder = TRUE)
+
+ds_test <- ds_test %>%
+  # see above reg. dataset_repeat
+  dataset_repeat() %>%
+  dataset_batch(batch_size)
+
+
+```
+
+With data manipulations as complicated as the above, it's always worthwhile checking some pipeline outputs. We can do that using the usual `reticulate::as_iterator` magic, provided that for this test run, we _don't_ disable V2 behavior. (Just restart the R session between a "pipeline checking" and the later modeling runs.)
+
+Here, in any case, would be the relevant code:
+
+
+```{r}
+# this piece needs TF 2 behavior enabled
+iter <- as_iterator(ds_test) # or any other dataset you want to check
+while (TRUE) {
+ item <- iter_next(iter)
+ if (is.null(item)) break
+ print(item)
+}
+```
+
+
+With that we're ready to create the model.
+
+## Model
+
+The model will be a rather simple convnet. The main difference between standard and DP training lies in the optimization procedure; thus, it's straightforward to first establish a non-DP baseline. Later, when switching to DP, we'll be able to reuse almost everything.
+
+Here, then, is the model definition valid for both cases:
+
+```{r}
+model <- keras_model_sequential() %>%
+  layer_conv_1d(
+      filters = 32,
+      kernel_size = 3,
+      activation = "relu"
+    ) %>%
+  layer_batch_normalization() %>%
+  layer_conv_1d(
+      filters = 64,
+      kernel_size = 5,
+      activation = "relu"
+    ) %>%
+  layer_batch_normalization() %>%
+  layer_conv_1d(
+      filters = 128,
+      kernel_size = 5,
+      activation = "relu"
+    ) %>%
+  layer_batch_normalization() %>%
+  layer_global_average_pooling_1d() %>%
+  layer_dense(units = 128, activation = "relu") %>%
+  layer_dense(units = 1)
+```
+
+We train the model with mean squared error loss. 
+
+```{r}
+optimizer <- optimizer_adam()
+model %>% compile(loss = "mse", optimizer = optimizer, metrics = metric_mean_absolute_error)
+
+num_epochs <- 20
+history <- model %>% fit(
+  ds_train, 
+  steps_per_epoch = n_train/batch_size,
+  validation_data = ds_test,
+  epochs = num_epochs,
+  validation_steps = n_test/batch_size)
+```
+
+
+## Baseline results
+
+
+After 20 epochs, mean absolute error is around 6 bpm: 
+
+```{r, eval=TRUE, echo=FALSE, layout="l-body-outset", fig.cap = "Training history without differential privacy."}
+knitr::include_graphics("images/baseline.png")
+```
+
+Just to put this in context, the MAE reported for subject S1 in the paper[@ReissISL19] -- based on a higher-capacity network, extensive hyperparameter tuning, and naturally, training on the complete dataset -- amounts to 8.45 bpm on average; so our setup seems to be sound.
```

Comment:
```suggestion
To put this in context, the MAE reported for subject S1 in the paper[@ReissISL19] -- based on a higher-capacity network, extensive hyperparameter tuning, and naturally, training on the complete dataset -- amounts to 8.45 bpm on average. So, our setup seems to be sound.
```

Reply:
I've been trying to get rid of "just"s in my writing of late (turns out I use it _a lot_).